### ISSUE TITLE
ValueError: Circular reference detected

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible/ansible#14671]
- Issue Type: Bug Report #11536 
- Ansible Version: ansible 2.0.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = /usr/share/ansible
- Ansible Configuration:
  gathering = explicit
  fact_caching = redis
  fact_caching_timeout = 86400
- Environment: Linux 4.2.0-27-generic #32-Ubuntu SMP Fri Jan 22 04:49:08 UTC 2016 x86_64 GNU/Linux
- Summary: No issue with the first play which is called in the role - Second play leads to the aforementioned issue
- Steps To Reproduce: see below

```
# san-francisco.yml
- hosts: spines-sf
  roles:
    - common
    - spine

# roles/common/tasks/main.yml
- name: get NX-OS facts
  connection: local
  nxos_get_facts: host={{ inventory_hostname }} username={{ username }} password={{ password }}
  tags: get-facts
- name: save NX-OS facts locally
  connection: local
  template: src=templates/nx-os_facts.j2 dest=files/facts/{{ inventory_hostname }}_nx-os_facts.json
  tags: save-facts

# templates/nx-os_facts.j2
DEVICE: {{ inventory_hostname }}
NX-OS Facts:
{{ nxos | to_nice_json }}

#./hosts with ANSIBLE_HOSTS=./hosts
[all:vars]
ansible_connection = local
username=admin
password=admin
[spines-sf]
NX-OSv-Spine-1[1:2]

# /etc/hosts
172.21.100.11 NX-OSv-Spine-11 
172.21.100.12 NX-OSv-Spine-12 
```
- Expected Results: No error
- Actual Results: 

```
ansible-playbook california.yml --limit spines-sf 

PLAY ***************************************************************************

TASK [common : get NX-OS facts] ************************************************
ok: [NX-OSv-Spine-12]
ok: [NX-OSv-Spine-11]

TASK [common : save NX-OS facts locally] ***************************************
fatal: [NX-OSv-Spine-11]: FAILED! => {"changed": false, "failed": true, "msg": "ValueError:*** Circular reference detected***"}
fatal: [NX-OSv-Spine-12]: FAILED! => {"changed": false, "failed": true, "msg": "ValueError: ***Circular reference detected***"}


```

**Comment by jean-christophe-manciot** [comment]
I have updated the previous post to take into account some recent changes in nxos_get_facts.

**Comment by jean-christophe-manciot** [comment]
I have found the cause of the issue: if I remove the first line (DEVICE: {{ inventory_hostname }}), the role gets executed correctly:

```
TASK [common : get NX-OS facts] ************************************************
ok: [NX-OSv-Spine-11]
ok: [NX-OSv-Spine-12]

TASK [common : save NX-OS facts locally] ***************************************
changed: [NX-OSv-Spine-11]
changed: [NX-OSv-Spine-12]
```

**Comment by bcoca** [comment]
seems like a weird interaction with delegating to the local.

**Comment by ashb** [comment]
Might be related to [#15610 "undefined variable + filter in module args casuses hard-to-track-down error(Regression from v1.9)"](https://github.com/ansible/ansible/issues/15610)

**Comment by ddohler** [comment]
I just encountered this error and it was indeed caused by #15610; I was receiving the error when using the `template` module. The template file that it was failing on contained only one variable that was being passed through a filter. After removing the filter, I received an `AnsibleUndefinedVariable` error related to a missing entry in one of my group_vars files. Once I resolved this, I replaced the filter in the template and everything worked.

**Comment by ozbillwang** [comment]
@ddohler 

thanks, it is really helpful to fix my issue.

**Comment by jimi-c** [comment]
Based on the above, and the fact that I just pushed a fix for #15610, I'm going to go ahead and close this now. This fix will also be included in the next release, 2.1.1.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
